### PR TITLE
DSND-3116: Modify behaviour of the delete endpoint

### PR DESCRIPTION
**Pull request by kinpang-CH**
* Refactor endpoint take in delta_at as a request header
* Refactor behaviour to compare on delta_at and check when its null
* Refactor API service and seperate transformation into new mapper class
* Refactor and use new ExceptionHandler class for controller
* Add new exception classes to be thrown instead of returning response entity
* Update unit and cucumber tests
* Passes local testing

Resolves [DSND-3116](https://companieshouse.atlassian.net/browse/DSND-3116)

[DSND-3116]: https://companieshouse.atlassian.net/browse/DSND-3116?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by ch-code-analysis** [comment]
CI: Security warnings found!

File: src/main/java/uk/gov/companieshouse/charges/data/config/ChargesApplicationConfig.java [hunk 1]
```diff
@@ -51,6 +55,22 @@ public InternalApiClient internalApiClient() {
         return ApiSdkManager.getPrivateSDK();
     }
 
+    @Bean
+    public Supplier<InternalApiClient> internalApiClientSupplier(
+            @Value("${chs.kafka.api.key}") String apiKey,
+            @Value("${chs.kafka.api.endpoint}") String apiUrl) {
+        return () -> {
+            InternalApiClient internalApiClient = new InternalApiClient(new ApiKeyHttpClient(apiKey));
```
Comment (sthompsonCH) [inline src/main/java/uk/gov/companieshouse/charges/data/config/ChargesApplicationConfig.java:63]: Can you add the request id to the ApiKeyHttpClient object here?

File: src/main/java/uk/gov/companieshouse/charges/data/api/ChargesApiService.java [hunk 2]
```diff
@@ -2,44 +2,67 @@
...
+    public ChargesApiService(Supplier<InternalApiClient> internalApiClientSupplier, ResourceChangedRequestMapper mapper) {
+        this.internalApiClientSupplier = internalApiClientSupplier;
+        this.mapper = mapper;
+    }
+
+
+    /**
+     * Calls the CHS Kafka api.
+     * @param resourceChangedRequest encapsulates details relating to the updated or deleted company exemption
+     * @return The service status of the response from chs kafka api
```
Comment (cgregoryCH) [inline src/main/java/uk/gov/companieshouse/charges/data/api/ChargesApiService.java:38]: This javadoc can be removed

File: src/main/java/uk/gov/companieshouse/charges/data/config/ChargesApplicationConfig.java (hunk 1 above)
Comment (kinpang-CH) [reply src/main/java/uk/gov/companieshouse/charges/data/config/ChargesApplicationConfig.java]: Resolved in latest commit

File: src/main/java/uk/gov/companieshouse/charges/data/api/ChargesApiService.java (hunk 2 above)
Comment (kinpang-CH) [reply src/main/java/uk/gov/companieshouse/charges/data/api/ChargesApiService.java]: Resolved in latest commit

**Comment by ch-code-analysis** [comment]
CI: Security warnings found!

**Comment by ch-code-analysis** [comment]
CI: Security warnings found!

**Review by cgregoryCH — APPROVED** [review]

**Review by sthompsonCH — APPROVED** [review]

**Pull request merged**